src/tiny_gea2_single_wire_interface.c

```diff
@@ -0,0 +1,647 @@
+{
+  if(instance->_private.receive.count == 0) {
+    instance->_private.receive.crc = tiny_gea3_crc_seed;
+  }
+
+  if(instance->_private.receive.count < instance->_private.receive.buffer_size) {
+    instance->_private.receive.buffer[instance->_private.receive.count++] = byte;
+
+    instance->_private.receive.crc = tiny_crc16_byte(
+      instance->_private.receive.crc,
+      byte);
+  }
+}
+
+static bool received_packet_is_addressed_to_me(self_t* self)
+{
+  reinterpret(packet, self->_private.receive.buffer, tiny_gea3_packet_t*);
+  return (packet->destination == self->_private.address) ||
+    is_broadcast_address(packet->destination) ||
+    self->_private.ignore_destination_address;
+}
+
+static void send_ack(self_t* instance, uint8_t address)
+{
+  if(!is_broadcast_address(address)) {
+    tiny_uart_send(instance->_private.uart, tiny_gea3_ack);
+  }
+}
+
+static void process_received_byte(self_t* instance, const uint8_t byte)
+{
+  reinterpret(packet, instance->_private.receive.buffer, tiny_gea3_packet_t*);
+
+  if(instance->_private.receive.escaped) {
+    instance->_private.receive.escaped = false;
+    buffer_received_byte(instance, byte);
+    return;
+  }
+
+  switch(byte) {
+    case tiny_gea3_esc:
+      instance->_private.receive.escaped = true;
+      break;
+
+    case tiny_gea3_stx:
+      instance->_private.receive.count = 0;
+      break;
+
+    case tiny_gea3_etx:
+      if(!received_packet_has_minimum_valid_length(instance) || !received_packet_has_valid_length(instance)) {
+        break;
+      }
+
+      if(!received_packet_has_valid_crc(instance)) {
+        break;
+      }
+
+      if(!received_packet_is_addressed_to_me(instance)) {
+        break;
+      }
+
+      packet->payload_length -= tiny_gea3_packet_transmission_overhead;
+      instance->_private.receive.packet_ready = true;
+
+      send_ack(instance, packet->destination);
+
+      tiny_fsm_transition(&instance->_private.fsm, state_idle_cooldown);
+      break;
+
+    default:
+      buffer_received_byte(instance, byte);
+      break;
+  }
+}
+
+static tiny_timer_ticks_t get_collision_timeout(uint8_t address, uint8_t pseudoRandomNumber)
+{
+  (void)pseudoRandomNumber;
+  return 43 + (address & 0x1F) + ((78 ^ address) & 0x1F);
+}
+
+static void collision_idle_timeout(void* context)
+{
+  self_t* instance = context;
+  tiny_fsm_send_signal(&instance->_private.fsm, signal_collision_idle_timeout, NULL);
+}
+
+static void start_collision_idle_timeout_timer(self_t* instance)
+{
+  uint8_t currentTicks = (uint8_t)tiny_time_source_ticks(instance->_private.timer_group.time_source);
+  tiny_timer_ticks_t collisionTimeoutTicks = get_collision_timeout(instance->_private.address, currentTicks);
+
+  tiny_timer_start(
+    &instance->_private.timer_group,
+    &instance->_private.timer,
+    collisionTimeoutTicks,
+    instance,
+    collision_idle_timeout);
+}
+
+static void state_collision_cooldown(tiny_fsm_t* fsm, const tiny_fsm_signal_t signal, const void* data)
```

Comment:
What are collisions?

Reply:
so this is a single wire protocol - there is one data line. If something tries to send a message at the same time as the other one, the messages "collide" and we need to wait for a cooldown period to finish before attempting to send again